### PR TITLE
prevent JS in link to github

### PR DESCRIPTION
**Pull request by lanwen**
This change is [Reviewable](https://reviewable.io/reviews/jenkinsci/github-plugin/114)

### CODE REVIEW

File: src/main/java/com/coravy/hudson/plugins/github/GithubLinkAction.java [hunk 1]
```diff
@@ -33,7 +34,7 @@ public String getIconFileName() {
 
     @Override
     public String getUrlName() {
-        return projectProperty.getProjectUrl().baseUrl();
+        return XSSApi.asValidHref(projectProperty.getProjectUrl().baseUrl());
```
Comment (KostyaSha) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAction.java]: `Functions.*escape*` `hudson.Util#escape`

File: src/main/java/com/coravy/hudson/plugins/github/GithubLinkAction.java [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
-import java.util.Collections;
-
 import hudson.Extension;
 import hudson.model.Action;
 import hudson.model.Job;
 import jenkins.model.TransientActionFactory;
+import org.jenkinsci.plugins.github.util.XSSApi;
+
+import java.util.Collection;
+import java.util.Collections;
```
Comment (KostyaSha) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAction.java]: Add checkstyle verification for imports?

File: src/main/java/org/jenkinsci/plugins/github/util/XSSApi.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package org.jenkinsci.plugins.github.util;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+/**
+ * @author lanwen (Merkushev Kirill)
+ */
+public final class XSSApi {
```
Comment (daniel-beck) [inline src/main/java/org/jenkinsci/plugins/github/util/XSSApi.java]: Maybe `@Restricted` if you care about plugins using APIs of this plugin?

File: src/main/java/org/jenkinsci/plugins/github/util/XSSApi.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    /**
+     * Method to filter invalid url for XSS. This url can be inserted to href safely
+     *
+     * @param urlString unsafe url
+     *
+     * @return safe url
+     */
+    public static String asValidHref(String urlString) {
+        try {
+            return new URL(urlString).toExternalForm();
```
Comment (daniel-beck) [inline src/main/java/org/jenkinsci/plugins/github/util/XSSApi.java]: What happens to relative links? Protocol-relative URLs?

File: src/main/java/com/coravy/hudson/plugins/github/GithubLinkAction.java (hunk 2 above)
Comment (lanwen) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAction.java]: Already is. What wrong now?

File: src/main/java/com/coravy/hudson/plugins/github/GithubLinkAction.java (hunk 1 above)
Comment (lanwen) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAction.java]: nope. They can't escape js

File: src/main/java/org/jenkinsci/plugins/github/util/XSSApi.java (hunk 4 above)
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/util/XSSApi.java]: Expecting github url is **scheme + fqdn + project path** all other urls will be filtered

File: src/main/java/com/coravy/hudson/plugins/github/GithubLinkAction.java (hunk 2 above)
Comment (KostyaSha) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAction.java]: Import order.

Comment (lanwen) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAction.java]: now it alphabetically ordered. As for me its ok

Comment (KostyaSha) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAction.java]: Yes, but re-ordering mean that it not defined in checkstyle. http://checkstyle.sourceforge.net/config_imports.html#ImportOrder_Example

Comment (lanwen) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAction.java]: this is not what we should care about

**Pull request merged**